src/taretto/modeling/__init__.py

```diff
@@ -0,0 +1,195 @@
+        if name not in self._availiable_collections:
+            sorted_collection_keys = sorted(self._availiable_collections)
+            raise AttributeError(
+                "Collection [{}] not known to object, available collections: {}".format(
+                    name, sorted_collection_keys
+                )
+            )
+        if name not in self._collection_cache:
+            item_filters = self._filters.copy()
+            cls_and_or_filter = self._availiable_collections[name]
+            if isinstance(cls_and_or_filter, tuple):
+                item_filters.update(cls_and_or_filter[1])
+                cls_or_verpick = cls_and_or_filter[0]
+            else:
+                cls_or_verpick = cls_and_or_filter
+            # Now check whether we verpick the collection or not
+            if isinstance(cls_or_verpick, VersionPick):
+                cls = cls_or_verpick.pick(self._parent.application.version)
+                try:
+                    # TODO Add logger
+                    print(
+                        "[COLLECTIONS] Version picked collection %s as %s.%s",
+                        name,
+                        cls.__module__,
+                        cls.__name__,
+                    )
+                except (AttributeError, TypeError, ValueError):
+                    # TODO Add logger
+                    print("[COLLECTIONS] Is the collection %s truly a collection?", name)
+            else:
+                cls = cls_or_verpick
+            self._collection_cache[name] = cls(self._parent, filters=item_filters)
+        return self._collection_cache[name]
+
+
+@attr.s
+class BaseCollection(object):
+    """Class for helping create consistent Collections
+
+    The BaseCollection class is responsible for ensuring two things:
+
+    1) That the API consistently has the first argument passed to it
+    2) That that first argument is an application instance
+
+    This class works in tandem with the entrypoint loader which ensures that the correct
+    argument names have been used.
+    """
+
+    ENTITY = None
+
+    parent = attr.ib(repr=False)
+    filters = attr.ib(default=attr.Factory(dict))
+
+    @property
+    def application(self):
+        if isinstance(self.parent, BaseEntity):
+            return self.parent.application
+        else:
+            return self.parent
+
+    @classmethod
+    def for_application(cls, application, *k, **kw):
+        return cls(application)
+
+    @classmethod
+    def for_entity(cls, obj, *k, **kw):
+        return cls(obj, *k, **kw)
+
+    @classmethod
+    def for_entity_with_filter(cls, obj, filt, *k, **kw):
+        return cls.for_entity(obj, *k, **kw).filter(filt)
+
+    def instantiate(self, *args, **kwargs):
+        return self.ENTITY.from_collection(self, *args, **kwargs)
+
+    def filter(self, filter):
+        filters = self.filters.copy()
+        filters.update(filter)
+        return attr.evolve(self, filters=filters)
```

Comment:
i suspect having a random dict as filters will set us up for pain

i'd like to see filters separated from collections to avoid the down payments that will eventually come from bringing them too close together